test/functional/test_framework/test_node.py

```diff
@@ -468,7 +468,10 @@ def add_p2p_connection(self, p2p_conn, *, wait_for_verack=True, **kwargs):
         p2p_conn.peer_connect(**kwargs, net=self.chain)()
         self.p2ps.append(p2p_conn)
         if wait_for_verack:
+            # Wait for the node to send us the version and verack
             p2p_conn.wait_for_verack()
+            # Force the node to fully process our verack as well
+            p2p_conn.sync_with_ping()
```

Comment:
It looks like this change would affect quite a few other functional tests that don't pass `wait_for_verack=False`. Is it a good idea to change it here rather than in p2p_blocksonly?

Reply:
Yes, whenever we add a p2p connection in the tests we want the least amount of surprise and want to assume that the connection has been fully established when the function returns.

Only in adversarial tests do we need to make sure that the connection is not (or only partially) established. Do you see any other test that needs a partial connection?

Reply:
There are some tests that now have a redundant `sync_with_ping()` immediately following an `add_p2p_connection()`, e.g. in p2p_invalid_messages.py::143

and others that have a single operation between the two calls and which are probably being needlessly synced twice (or perhaps should not be synced until after the single operation), e.g. in 3 places in p2p_dos_header_tree.py

Reply:
Not sure, but maybe an incremental change would be safer.

Reply:
I think this is fine, at worst slows down the test by a few milliseconds.

> There are some tests that now have a redundant sync_with_ping() immediately following an add_p2p_connection(), e.g. in p2p_invalid_messages.py::143

I fixed this instance in the latest commit, but couldn't find any other instances.

> and others that have a single operation between the two calls and which are probably being needlessly synced twice (or perhaps should not be synced until after the single operation), e.g. in 3 places in p2p_dos_header_tree.py

Why are they needlessly being synced? First they need to wait until the connection is fully established, then they need to wait until the message they sent is fully processed. This is a common pattern in our tests, see `send_and_ping`.

I went ahead and replaced the two-line calls there with the more widely used one-line `send_and_ping`

Reply:
You're welcome :)

> Why are they needlessly being synced? 

Unless I am confused, these cases seem to be synced one additional time after this change:
- Before: add connection / send / sync
- After: add connection including sync by default / send / sync

Help me understand if that's not the case.

Reply:
Ok fine, in this case the sync is redundant, but I can't see a way to distinguish the redundant case from the non-redundant one. Even if we could, the cost of a few milliseconds is probably not worth it to optimize this, right?

Reply:
Force pushed with additional doc to explain my reasoning

Reply:
I agree; thanks for the doc